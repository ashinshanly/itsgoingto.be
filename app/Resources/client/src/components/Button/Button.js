import React from 'react'
import PropTypes from 'prop-types'
import EventBus from 'components/EventBus'
import Spinner from 'components/Spinner/Spinner'
import './Button.scss'

class Button extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      disabled : false,
      loading  : false
    }
  }

  componentDidMount = () => {
    this._mounted = true
    this.eventBus = EventBus.getEventBus()
    if (this.props.submitEvent) {
      this.eventListener = this.eventBus.addListener(this.props.submitEvent, () => {
        this.handlePress()
      })
    }
  }

  componentWillUnmount = () => {
    this._mounted = false
    if (this.eventListener) {
      this.eventListener.remove()
    }
  }

  isDisabled = () => this.props.disabled || this.state.disabled

  handlePress = (event) => {
    if (event) {
      event.preventDefault()
    }
    if (!this.isDisabled()) {
      this.setState({
        disabled : true,
        loading  : true
      })
      this.props.callback().then((reset) => {
        if (reset !== false && this._mounted) {
          this.setState({
            disabled : false,
            loading  : false
          })
        }
      })
    }
  }

  render = () => (
    <button
      className={'btn ' + this.props.className + (this.isDisabled() ? ' disabled' : '')}
      disabled={this.props.disabled}
      onClick={this.handlePress}>
      {this.state.loading && <Spinner />}
      {!this.state.loading && <span>{this.props.text}</span>}
    </button>
  )
}

Button.propTypes = {
  text        : PropTypes.string.isRequired,
  className   : PropTypes.string,
  disabled    : PropTypes.bool,
  callback    : PropTypes.func.isRequired,
  submitEvent : PropTypes.string
}

Button.defaultProps = {
  disabled : false,
}

export default Button
