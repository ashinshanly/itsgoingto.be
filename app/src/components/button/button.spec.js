/* global expect, jest */
import React from 'react'
import { shallow } from 'enzyme'
import Button from './button'
import EventBus from 'components/event-bus'

describe('(Component) Button', () => {
  const wrapper = shallow(<Button />)
  const instance = wrapper.instance()

  it('renders as a button', () => {
    expect(instance).toBeInstanceOf(Button)
    expect(wrapper.name()).toBe('button')
  })

  describe('(State)', () => {
    it('Should have initial state', () => {
      expect(wrapper.state()).toEqual({ disabled : false, loading : false })
    })
  })

  describe('(Props)', () => {
    it('Should have default props', () => {
      expect(instance.props.text).toBe('')
      expect(instance.props.className).toBe('')
      expect(instance.props.disabled).toBe(false)
      expect(instance.props.callback).toBe(null)
      expect(instance.props.submitEvent).toBe(null)
    })
  })

  describe('(Method) isDisabled', () => {
    it('Should return false as default', () => {
      const wrapper = shallow(<Button />).instance()

      expect(wrapper.isDisabled()).toBe(false)
    })

    it('Should return true for prop', () => {
      const wrapper = shallow(<Button disabled />).instance()

      expect(wrapper.isDisabled()).toBe(true)
    })

    it('Should return true for state', () => {
      const wrapper = shallow(<Button />).instance()

      wrapper.setState({ disabled: true })
      expect(wrapper.isDisabled()).toBe(true)
    })
  })

  describe('(Method) handlePress', () => {
    it('Should update the state', () => {
      const wrapper = shallow(<Button callback={() => Promise.resolve()} />)
      const instance = wrapper.instance()

      instance.handlePress()
      expect(wrapper.state()).toEqual({ disabled : true, loading : true })
    })

    it('Should call callback', () => {
      const callback = jest.fn(() => (Promise.resolve()))
      const wrapper = shallow(<Button callback={callback} />)

      wrapper.instance().handlePress()
      expect(callback).toHaveBeenCalledTimes(1)
    })

    it('Shouldn\'t update the state when disabled', () => {
      const wrapper = shallow(<Button disabled callback={() => Promise.resolve()} />)
      const instance = wrapper.instance()

      instance.handlePress()
      expect(wrapper.state()).toEqual({ disabled : false, loading : false })
    })

    it('Shouldn\'t update the state with no callback', () => {
      const wrapper = shallow(<Button />)
      const instance = wrapper.instance()

      instance.handlePress()
      expect(wrapper.state()).toEqual({ disabled : false, loading : false })
    })

    // Note : I don't seem to be able to test the promise callback in handlePress
  })

  describe('(EventListener) submitEvent', () => {
    const wrapper = shallow(<Button submitEvent='submitButton' />)
    const instance = wrapper.instance()

    instance.handlePress = jest.fn()

    it('Should call handlePress', () => {
      EventBus.getEventBus().emit('submitButton')
      expect(instance.handlePress).toHaveBeenCalledTimes(1)
    })
  })

  describe('(Render) snapshot', () => {
    it('Should render the text', () => {
      const wrapper = shallow(<Button text='Click Me!' />)

      expect(wrapper).toMatchSnapshot()
    })

    it('Should be disabled', () => {
      const wrapper = shallow(<Button text='Click Me!' disabled />)

      expect(wrapper).toMatchSnapshot()
    })

    it('Should have className', () => {
      const wrapper = shallow(<Button className='button button--class' disabled />)

      expect(wrapper).toMatchSnapshot()
    })

    it('Should render spinner when loading', () => {
      const wrapper = shallow(<Button text='Click Me!' />)

      wrapper.setState({ disabled : true, loading : true })
      expect(wrapper).toMatchSnapshot()
    })
  })
})
