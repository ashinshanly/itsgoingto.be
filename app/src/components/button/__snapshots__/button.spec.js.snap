// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) Button (Render) snapshot Should be disabled 1`] = `
<button
  className="btn  disabled"
  disabled={true}
  onClick={[Function]}
>
  <span>
    Click Me!
  </span>
</button>
`;

exports[`(Component) Button (Render) snapshot Should have className 1`] = `
<button
  className="btn button button--class disabled"
  disabled={true}
  onClick={[Function]}
>
  <span />
</button>
`;

exports[`(Component) Button (Render) snapshot Should render spinner when loading 1`] = `
<button
  className="btn  disabled"
  disabled={false}
  onClick={[Function]}
>
  <Spinner />
</button>
`;

exports[`(Component) Button (Render) snapshot Should render the text 1`] = `
<button
  className="btn "
  disabled={false}
  onClick={[Function]}
>
  <span>
    Click Me!
  </span>
</button>
`;
